### PR TITLE
Add more options for WinRM

### PR DESCRIPTION
**Pull request by smurawski**
Uses winrm-s to allow sspinegotiate auth from Windows hosts to Windows guests.

Adds winrm_transport to the configuration options for the winrm transport with defaults to plaintext (the current behavior) and can be set to sspinegotiate.

The effect is that we don't have to have AllowUnencypted set to $true for the WinRM service on the target machines.

This won't help drivers that have their own reliance on WinRM as a communication option (like kitchen-vagrant), but any driver that uses the transport abstraction can take advantage of this.

### CODE REVIEW

**Comment by smurawski** [comment]
@tyler-ball @fnichol test failure looks unrelated to my change.  Thoughts?

**Comment by ksubrama** [comment]
Hey, I don't see a winrm-s dependency anywhere in the gemspec (and I only see a winrm development dependency there).  How does winrm-s get pulled in?

**Comment by zl4bv** [comment]
It looks like winrm-s is bundled with chefdk, but that's no help if you're running test-kitchen some other way.

File: lib/kitchen/transport/winrm.rb [hunk 1]
```diff
@@ -93,14 +104,14 @@ def execute(command)
             log_stderr_on_warn(stderr)
           elsif exit_code != 0
             log_stderr_on_warn(stderr)
-            raise Transport::WinrmFailed,
-              "WinRM exited (#{exit_code}) for command: [#{command}]"
+            fail Transport::WinrmFailed,
```
Comment (tyler-ball) [inline lib/kitchen/transport/winrm.rb]: Why this change?

File: lib/kitchen/transport/winrm.rb [hunk 2]
```diff
@@ -122,9 +133,9 @@ def upload(locals, remote)
         def wait_until_ready
           delay = 3
           session(
-            :retries  => max_wait_until_ready / delay,
-            :delay    => delay,
-            :message  => "Waiting for WinRM service on #{endpoint}, " \
+            retries: max_wait_until_ready / delay,
```
Comment (tyler-ball) [inline lib/kitchen/transport/winrm.rb]: I actually don't think we'll be able to make this change - some people still use TK with Ruby 1.9

**Comment by tyler-ball** [comment]
I'm generally :+1: on this.

I'm not sure what the test failures are...  Can you rebase off the latest master?

**Comment by smurawski** [comment]
@tyler-ball Yeah, the test failures is what has sent me down the route of making the tests run on Windows, so this is blocked on https://github.com/test-kitchen/test-kitchen/pull/689

I'm also adding SSL support, since we are enabling ssl for winrm in knife and provisioning.

File: lib/kitchen/transport/winrm.rb (hunk 1 above)
Comment (smurawski) [inline lib/kitchen/transport/winrm.rb]: I don't remember, I think rubocop made that change or suggested it.. I'll revert it.

File: lib/kitchen/transport/winrm.rb (hunk 2 above)
Comment (smurawski) [inline lib/kitchen/transport/winrm.rb]: I can flip the syntax back, but due to the aruba update, we can only test on 2.x and newer.

File: lib/kitchen/transport/winrm.rb [hunk 3]
```diff
@@ -38,22 +36,36 @@ class WinrmFailed < TransportFailed; end
...
       # (see Base#connection)
       def connection(state, &block)
+        case config[:winrm_transport]
+        when "ssl"
+          config[:winrm_transport] = :ssl
+        when "sspinegotiate" && host_os_windows?
+          config[:winrm_transport] = :sspinegotiate
+        else
+          config[:winrm_transport] = :plaintext
+        end
```
Comment (fnichol) [inline lib/kitchen/transport/winrm.rb]: This block would work better in a `def finalize_config!(instance)` method, similar to https://github.com/test-kitchen/test-kitchen/blob/621cd45784c3aed0e893d390c2758125b250151a/lib/kitchen/transport/ssh.rb#L62-L75 (note the `super` call there is pretty important :smile:)

The reason is that `kitchen diagnose` returns its configuration after the `#finalize_config!` has done any last minute tweaks, meaning that the final computed value of `config[:winrm_transport]` would show in the diagnostic output.

File: lib/kitchen/transport/winrm.rb [hunk 4]
```diff
@@ -452,8 +477,29 @@ def load_needed_dependencies! # rubocop:disable Metrics/MethodLength
...
+        case RbConfig::CONFIG["host_os"]
+        when /darwin/
+          false
+        when /mswin|msys|mingw|cygwin|bccwin|wince|emc/
+          true
+        when /linux/
+          false
+        else
+          false
+        end
```
Comment (fnichol) [inline lib/kitchen/transport/winrm.rb]: I'd say you're okay with the super simple case of:

```ruby
case RbConfig::CONFIG["host_os"]
when /mswin|msys|mingw|cygwin|bccwin|wince|emc/
  true
else
  false
end
```

That way we're focused on the whilelisting tokens and dropping anything else by default.

Comment (fnichol) [inline lib/kitchen/transport/winrm.rb]: Now I see the code block this code was taken from :)

File: lib/kitchen/transport/winrm.rb [hunk 5]
```diff
@@ -452,8 +477,29 @@ def load_needed_dependencies! # rubocop:disable Metrics/MethodLength
...
+      def sspi_enabled?
+        enabled = host_os_windows? &&
+          config[:winrm_transport] == :sspinegotiate
+        logger.debug("Checking for Negotiate Auth: Enabled = #{enabled}")
+        enabled
+      end
+
       def load_winrm_transport!
         silence_warnings { require "winrm" }
+        silence_warnings { require "winrm-s" } if sspi_enabled?
```
Comment (fnichol) [inline lib/kitchen/transport/winrm.rb]: I think we're going to need a little load/rescue code for this gem, similar in spirit to [#load_needed_dependencies!](https://github.com/test-kitchen/test-kitchen/blob/b73b8fbccaf84242efb3c05502e36557c963c83a/lib/kitchen/transport/winrm.rb#L426-L450) to handle the case where we want the gem, try to require it, but the gem can't be activated (primarily this for outside ChefDK, but still a valid failure case). I was trying to see where best to handle this, and am currently thinking a short method with the require and `rescue LoadError` blocks? Happy to chat more about this one--I think it's the first case we have of augmenting behavior with an addition gem. Can't think of another place to look for stealing ideas!

**Comment by fnichol** [comment]
Other the inline comments above, any reason we can't add some tests to cover the data-in/config out behavior? Happy to help with ideas, direction, pairing, or code. I'm pretty sure we (finally) have enough examples to draw on this seems doable, woo!

**Comment by smurawski** [comment]
@fnichol @tyler-ball @randomcamel It took me a bit to get back to this, but I've got the code refactored and restructured to address the concerns mentioned above.  I've also added tests to cover the configuration options and conditional loading of winrm-s.

**Comment by andrewelizondo** [comment]
Any update on getting this merged? @fnichol @tyler-ball @randomcamel

File: lib/kitchen/transport/winrm.rb [hunk 6]
```diff
@@ -38,19 +36,43 @@ class WinrmFailed < TransportFailed; end
...
 
-      default_config :port, 5985
       default_config :username, "administrator"
       default_config :password, nil
-      default_config :endpoint_template, "http://%{hostname}:%{port}/wsman"
       default_config :rdp_port, 3389
       default_config :connection_retries, 5
       default_config :connection_retry_sleep, 1
       default_config :max_wait_until_ready, 600
+      default_config :winrm_transport, "plaintext"
```
Comment (mwrock) [inline lib/kitchen/transport/winrm.rb]: default to sspinegotiate on windows hosts

File: lib/kitchen/transport/winrm.rb [hunk 7]
```diff
@@ -389,24 +410,38 @@ def to_s
...
+          :winrm_transport => data[:winrm_transport]
         }
-
+        opts.merge!(additional_transport_args(opts[:winrm_transport]))
         opts
       end
 
+      def additional_transport_args(transport_type)
+        case transport_type
+        when :ssl, :sspinegotiate
```
Comment (mwrock) [inline lib/kitchen/transport/winrm.rb]: I suggest we match the logic in knife-windows 1.x:
`:disable_sspi` if not on windows, using `ssl, or not not authenticating with `negotiate`
`:basic_auth_only` only from linux when ssl is not specified

File: lib/kitchen/transport/winrm.rb [hunk 8]
```diff
@@ -424,41 +459,90 @@ def create_new_connection(options, &block)
...
-        end
-      rescue LoadError => e
-        logger.fatal("The `winrm-transport' gem is missing and must" \
-          " be installed or cannot be properly activated. Run" \
-          " `gem install winrm-transport --version '#{spec_version}'`" \
-          " or add the following to your Gemfile if you are using Bundler:" \
-          " `gem 'winrm-transport', '#{spec_version}'`.")
-        raise UserError,
-          "Could not load or activate WinRM::Transport (#{e.message})"
+        load_winrm_s! if host_os_windows?
```
Comment (mwrock) [inline lib/kitchen/transport/winrm.rb]: We probably should not load winrm-s if using `ssl`

**Comment by mwrock** [comment]
I'm totally :+1: on adding these winrm options but we should try to mach the defaults currently in use by knife-windows

**Comment by smurawski** [comment]
@mwrock I'm open to matching the defaults in knife-windows, but that'll be a breaking change. This PR keeps the same default behavior but supports additional options which I think is better for a minor release.

**Comment by mwrock** [comment]
Yeah I agree we don't want this to be a breaking change. I wasn't thinking that would be breaking though. Which scenario does this break?

**Comment by smurawski** [comment]
If the defaults were changed to the same as knife-windows, any cross-platform connections would fail out of the box (since all previous kitchen connections would have been basic/unencrypted).  From Windows, we'd be ok because winrm-s.

**Comment by mwrock** [comment]
Ah right we throw an error now asking the user to be explicit about basic auth. So I'd say keep the knife-windows strategy EXCEPT for linux hosts default to `basic_auth`

**Comment by smurawski** [comment]
@tyler-ball This PR should be good to go.  The test failures are the same as the test failures against master.
